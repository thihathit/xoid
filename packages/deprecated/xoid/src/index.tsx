export type { Init, Atom, Listener, StateOf, GetState } from '@xoid/engine'
export type { Usable, Enhancer } from './core'

export { subscribe, effect } from '@xoid/engine'
export { create, use } from './core'
