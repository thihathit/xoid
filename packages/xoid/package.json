{
  "name": "xoid",
  "version": "1.0.0-beta.0",
  "main": "./index.js",
  "module": "./index.esm.js",
  "types": "./index.d.ts",
  "sideEffects": false,
  "description": "Framework-agnostic state management library designed for simplicity and scalability",
  "homepage": "https://xoid.dev",
  "repository": {
    "type": "git",
    "url": "https://github.com/onurkerimov/xoid",
    "directory": "./packages/xoid"
  },
  "license": "MIT",
  "keywords": [
    "framework-agnostic",
    "state",
    "management",
    "state-management",
    "react",
    "redux",
    "recoil",
    "xoid",
    "atom",
    "store",
    "stream",
    "observable"
  ]
}