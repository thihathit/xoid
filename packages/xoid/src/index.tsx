import type { Atom, Stream, Init, GetState, Usable } from './types'
import { createSelector, createApi, INTERNAL } from './internal/utils'
import { createEvent, createInternal } from './internal/lite'

export * from './types'

/**
 * Creates an atom with the first argument as the initial state.
 * Second argument can be used to attach more actions to the atom.
 * @see [xoid.dev/docs/api/create](https://xoid.dev/docs/api/create)
 */
export function create<T>(): Stream<T>
export function create<T>(init: Init<T>): Atom<T>
export function create<T>(init: Init<T>, getUsable: null): Atom<T>
export function create<T, U>(init: Init<T>, getUsable?: (atom: Atom<T>) => U): Atom<T> & Usable<U>
export function create<T, U = undefined>(
  init?: Init<T>,
  getUsable?: null | ((atom: Atom<T>) => U)
) {
  const isFunction = typeof init === 'function'
  const initialValue = (isFunction ? undefined : init) as T
  const internal = createInternal(initialValue, (() => (use as any).devtools.send(atom)) as any)
  if (isFunction) internal.get = createSelector(internal, init as (get: GetState) => T)
  const atom = createApi(internal)
  internal.usable = getUsable?.(atom)
  // devtools support
  Object.defineProperty(atom, 'debugValue', {
    set(debugValue: string) {
      internal.debugValue = debugValue
      ;(use as any).devtools.init(atom, initialValue, isFunction)
    },
  })
  return atom
}

/**
 * Gets the "usables" of an atom.
 * @see [xoid.dev/docs/api/use](https://xoid.dev/docs/api/use)
 */
export const use = <T extends any>(atom: Usable<T>): T =>
  (use as any).devtools.wrap((atom as any)[INTERNAL].usable, atom)
;(use as any).symbol = INTERNAL
;(use as any).createEvent = createEvent
;(use as any).devtools = {
  init: <T,>(_atom: T) => void 0,
  send: <T,>(_atom: T) => void 0,
  wrap: <T,>(value: T): T => value,
}
