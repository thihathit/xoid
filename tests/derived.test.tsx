// @ts-ignore
import { create, use, subscribe, Atom } from 'xoid'

const debug = (atom: Atom<any>) => {
  return {
    self: atom,
    selfSerialized: JSON.stringify(atom),
    get: atom(),
    getSerialized: JSON.stringify(atom()),
    use: use(atom as any),
  }
}

it('creates a derived atom from multiple atoms', () => {
  const a = create(3)
  const b = create(5)
  const atom = create((get) => get(a) + get(b))
  expect(debug(atom)).toMatchSnapshot()
})

it('creates a derived atom using the same atom', () => {
  const firstAtom = create({ alpha: 3, beta: 5 })
  const atom = create((get) => {
    const value = get(firstAtom)
    return value.alpha + value.beta
  })
  expect(debug(atom)).toMatchSnapshot()
})

it('creates a derived atom using the same atom using selectors', () => {
  const firstAtom = create({ alpha: 3, beta: 5 })
  const atom = create((get) => get(firstAtom, 'alpha') + get(firstAtom, 'beta'))
  expect(debug(atom)).toMatchSnapshot()
})

it('creates a derived atom from multiple atoms (keeps in sync)', () => {
  const a = create(3)
  const b = create(5)
  const atom = create((get) => get(a) + get(b))
  expect(atom()).toBe(8)

  a((s) => s + 1)
  expect(atom()).toBe(9)
})

it('creates a derived atom using the same atom (keeps in sync)', () => {
  const firstAtom = create({ alpha: 3, beta: 5 })
  const atom = create((get) => {
    const value = get(firstAtom)
    return value.alpha + value.beta
  })
  expect(atom()).toBe(8)

  use(firstAtom, 'alpha')((s) => s + 1)
  expect(atom()).toBe(9)
})

it('creates a derived atom using the same atom using selectors (keeps in sync)', () => {
  const firstAtom = create({ alpha: 3, beta: 5 })
  const atom = create((get) => get(firstAtom, 'alpha') + get(firstAtom, 'beta'))
  expect(atom()).toBe(8)

  use(firstAtom, 'alpha')((s) => s + 1)
  expect(atom()).toBe(9)
})

test('can derive state from external sources', () => {
  const fakeReduxBase = create(8)
  const fakeRedux = {
    getState: () => fakeReduxBase(),
    subscribe: (item: any) => subscribe(fakeReduxBase, item),
  }

  const reduxAtom = create((get) => get(fakeRedux.getState, fakeRedux.subscribe))
  const listener = jest.fn()

  expect(reduxAtom()).toBe(8)

  subscribe(reduxAtom, listener)
  expect(listener).not.toBeCalled()
  reduxAtom((s) => s + 1)
  expect(listener).toBeCalled()
  expect(listener).toBeCalledWith(9, 8)
})
