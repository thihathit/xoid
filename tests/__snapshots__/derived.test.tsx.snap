// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a derived atom from multiple atoms 1`] = `
Object {
  "get": 8,
  "getSerialized": "8",
  "self": Object {
    "focus": [Function],
    "map": [Function],
    "set": [Function],
    "subscribe": [Function],
    "update": [Function],
    "value": 8,
    "watch": [Function],
    Symbol(): Object {
      "evaluate": [Function],
      "get": [Function],
      "listeners": Set {},
      "set": [Function],
      "subscribe": [Function],
      "usable": undefined,
    },
  },
  "selfSerialized": "{\\"value\\":8}",
  "use": undefined,
}
`;

exports[`creates a derived atom using the same atom 1`] = `
Object {
  "get": 8,
  "getSerialized": "8",
  "self": Object {
    "focus": [Function],
    "map": [Function],
    "set": [Function],
    "subscribe": [Function],
    "update": [Function],
    "value": 8,
    "watch": [Function],
    Symbol(): Object {
      "evaluate": [Function],
      "get": [Function],
      "listeners": Set {},
      "set": [Function],
      "subscribe": [Function],
      "usable": undefined,
    },
  },
  "selfSerialized": "{\\"value\\":8}",
  "use": undefined,
}
`;

exports[`creates a derived atom using the same atom using selectors 1`] = `
Object {
  "get": 8,
  "getSerialized": "8",
  "self": Object {
    "focus": [Function],
    "map": [Function],
    "set": [Function],
    "subscribe": [Function],
    "update": [Function],
    "value": 8,
    "watch": [Function],
    Symbol(): Object {
      "evaluate": [Function],
      "get": [Function],
      "listeners": Set {},
      "set": [Function],
      "subscribe": [Function],
      "usable": undefined,
    },
  },
  "selfSerialized": "{\\"value\\":8}",
  "use": undefined,
}
`;
